### PR TITLE
Lesson 24

### PR DESCRIPTION
**Pull request by MichaelTim75**
95% on methods (19/20)
93% on lines

### CODE REVIEW

File: src/test/java/edu/hillel/lesson24/DBStorageTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    private Storage storage;
+
+    @Before
+    public void setUp() {
+        storage = storageFactory.createStorage(StorageType.DB);
+        storage.clearData();
+    }
+
+    @Test
+    public void dataTest() {
```
Comment (kartius) [inline src/test/java/edu/hillel/lesson24/DBStorageTest.java:24]: How can I build this test if I don't have db server locally? Or how Jenkins builds the application if db server doesn't exist?

File: src/test/java/edu/hillel/lesson24/StorageFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    private StorageFactory storageFactory;
+
+    @Before
+    public void setUp() throws Exception {
+        storageFactory = new StorageFactory();
+    }
+
+    @Test
+    public void createStorage() {
+        Storage storage = storageFactory.createStorage(StorageType.IN_MEMORY);
```
Comment (kartius) [inline src/test/java/edu/hillel/lesson24/StorageFactoryTest.java:20]: What about other types of storages?

File: src/test/java/edu/hillel/lesson24/DBStorageTest.java (hunk 1 above)
Comment (MichaelTim75) [reply src/test/java/edu/hillel/lesson24/DBStorageTest.java:24]: I used docker and docker-compose to start db. Docker-compose also attached to the project& Init sql also.
As I understand yesterday homework - it's about testing, not about full infrastucture. 
Yes - we could use testcontainers, spring-boot, liquibase and so-on.
But I think our excercise was about testing functionality.
Also - we could have texting enviroment with test-db. And as local tests as jenkins could run tests on it.
I don't understand what is incorrect.......

File: src/test/java/edu/hillel/lesson24/StorageFactoryTest.java (hunk 2 above)
Comment (MichaelTim75) [reply src/test/java/edu/hillel/lesson24/StorageFactoryTest.java:20]: I test method createStorage - if I set storage type=A - I get storage type of A class.
Yes - I can add tests for all types - but I don't think that it's needed.
Testing of correct work of all storage types exists in each classes.
Test of method storageFactory.createStorage - I've done - one test for correct and and one for exception.

Comment (kartius) [reply src/test/java/edu/hillel/lesson24/StorageFactoryTest.java:20]: What is somebody changes internal implementation inside createStorage? For example, change creating a DB storage?

Comment (MichaelTim75) [reply src/test/java/edu/hillel/lesson24/StorageFactoryTest.java:20]: I understand you, but count of types can be very big and count of differences can also be very big.
For example - calculator and operation sum (a+b)
We make test that check than 1+2=3
But in method - smebody can change it implementation and do so
if (a==10){
return a*15-b+8;
}
else{
return a+b;
}
So - we don't cover all issues....
That's why I think that we check correctness of method, but not all possible variant of method executions.

Comment (kartius) [reply src/test/java/edu/hillel/lesson24/StorageFactoryTest.java:20]: Anyway, you need to try to cover all cases in your code, at least by happy path
Here you try to commit untested part of code and it's a wrong way

File: src/test/java/edu/hillel/lesson24/DBStorageTest.java (hunk 1 above)
Comment (kartius) [reply src/test/java/edu/hillel/lesson24/DBStorageTest.java:24]: Okey, why you didn't commit your docker-compose file? 
There are two ways to deal with it
First one - manual, you will commit your docker-compose file and add readme with description how to run tests correctly.
Second one - automatic, you will add a maven docker-compose plugin which will be automatically up and down docker-compose with db during a tests run

File: src/test/java/edu/hillel/lesson24/StorageFactoryTest.java (hunk 2 above)
Comment (MichaelTim75) [reply src/test/java/edu/hillel/lesson24/StorageFactoryTest.java:20]: Ok.
But sorry - I don't agree that this code untested - it tested in DBStorageTest and FileStorageTest.
Also idea shows that all methods are covered with test.
But ok - I'll add tests for this variants of usage.

File: src/test/java/edu/hillel/lesson24/DBStorageTest.java (hunk 1 above)
Comment (kartius) [reply src/test/java/edu/hillel/lesson24/DBStorageTest.java:24]: Screenshot 2023-10-31 at 16 29 38

This it PR which I can see

**Pull request closed without merging**